src/sympc/session/session.py

```diff
@@ -186,44 +186,53 @@ def get_protocol(self) -> Protocol:
         return self.protocol
 
     def _generate_random_share(
-        self,
-        shape: Union[tuple, torch.Size],
+        self, shape: Union[tuple, torch.Size], ring_size: int
     ) -> List[torch.Tensor]:
-        """Generate random tensor share for the given shape.
+        """Generate random tensor share for the given shape and ring_size.
 
         Args:
             shape (Union[tuple, torch.Size]): Shape for the share.
+            ring_size (int): ring size to generate share.
 
         The generators are invoked in Counter(CTR) mode as parties with the same initial seeds
-        could generate the same random number on subsequent invocations.
+        could generate correlated random numbers on subsequent invocations.
 
         Returns:
             List[torch.Tensor] : shares generated by the generators.
         """
+        from sympc.tensor import PRIME_NUMBER
+
         gen1, gen2 = self.przs_generators
+        tensor_type = get_type_from_ring(ring_size)
 
         tensor_share1 = generate_random_element(
-            tensor_type=self.tensor_type,
+            tensor_type=tensor_type,
             generator=gen1,
             shape=shape,
         )
 
         tensor_share2 = generate_random_element(
-            tensor_type=self.tensor_type,
+            tensor_type=tensor_type,
             generator=gen2,
             shape=shape,
         )
 
-        return [tensor_share1, tensor_share2]
+        # PRIME_NUMBER ring share is generated in torch.uint8 datatype
+        if ring_size == PRIME_NUMBER:
+            return [tensor_share1 % ring_size, tensor_share2 % ring_size]
```

Comment:
I am not sure If doing this generates uniformly random numbers between 0 and 67.
For example, if we generate random numbers between 0 and 4 (inclusive) and this is the `tensor_share` and now we want to generate random numbers between 0 and 1 we would do a `module` with 2.

In this case, we would end up with a number from `(random number from [0, 1, 2, 3, 4]) % 2`, but if you take a closer look you see that we have 3 even numbers (0, 2, 4) and 2 odd numbers (1 and 3) -- so we get numbers in [0, 1] but we tend to get 0 more often than 1.

I am not sure if it is the same case here.

**LE**: Thinking more into this, it might not affect us that much, but we still need to take into consideration if this is really uniformly distributed (Ping me on slack if you have questions)

Reply:
Good catch george , I will work on it :+1:  ,

Reply:
Generating non uniform random numbers,is a big security flaw, I thought to inspect at it after initial devlopment, a very good catch george, I have fixed the same setting max value in torchcsprng.